### PR TITLE
Update promql query of cpu and memory metrics for sig-performance tests

### PR DESCRIPTION
**Pull request by Sreeja1725**
### What this PR does
Before this PR:

- VirtControllerCPUUsage collected the avg usage.

- minVirtAPIMemUsageInMB and maxVirtAPIMemUsageInMB these metrics resulted in the single lowest or highest observed memory usage for each container in the given time frame. Due to this the weekly average graph fluctuated significantly, appearing inconsistent and "noisy".

- minVirtControllerMemUsageInMB and maxVirtControllerMemUsageInMB calculate the minimum or maximum usage for each pod and then took the maximum value.

After this PR:

- VirtControllerCPUUsage metric will take into consideration of the leader pod.

- updated minVirtAPIMemUsageInMB and maxVirtAPIMemUsageInMB to return average memory usage for all the containers in the given time frame.By averaging the query smoothens the data.

- minVirtControllerMemUsageInMB and maxVirtControllerMemUsageInMB  directly calculates the maximum of (min or max) usage without grouping reducing the complexity of the query and producing a single and stable value.

Fixes #

### Why we need it and why it was done in this way
These updates ensure more stable graphs for analyzing the memory and CPU usage trends for virt-api and virt-controller pods.
### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
Update promql query of cpu and memory metrics for sig-performance tests
```

### CODE REVIEW

File: tools/perfscale-audit/metric-client/metric-client.go [hunk 1]
```diff
@@ -54,7 +54,7 @@ const (
 	avgVirtAPIMemUsageInMB        = `avg(avg_over_time(container_memory_rss{pod=~"virt-api.*", container!="POD", container!=""}[%ds]))/1024/1024`
 	minVirtAPIMemUsageInMB        = `min_over_time(container_memory_rss{pod=~"virt-api.*", container!="POD", container!=""}[%ds])/1024/1024`
 	maxVirtAPIMemUsageInMB        = `max_over_time(container_memory_rss{pod=~"virt-api.*", container!="POD", container!=""}[%ds])/1024/1024`
-	avgVirtControllerCPUUsage     = `avg(sum(rate(container_cpu_usage_seconds_total{namespace="kubevirt",pod=~"virt-controller.*", container!="",container!="POD"}[%ds])) by (pod))`
+	VirtControllerCPUUsage        = `max(sum(rate(container_cpu_usage_seconds_total{namespace="kubevirt",pod=~"virt-controller.*", container!="",container!="POD"}[%ds])) by (pod))`
```
Comment (alaypatel07) [inline tools/perfscale-audit/metric-client/metric-client.go:57]: nit: I think we can get rid of the sum, 
```
max(rate(container_cpu_usage_seconds_total{namespace="kubevirt",pod=~"virt-controller.*", container!="",container!="POD"}[%ds])) by (pod)
```
will have the same effect

File: tools/perfscale-audit/metric-client/metric-client.go [hunk 2]
```diff
@@ -50,15 +50,15 @@ const (
 //	at all times.
 const (
 	vmiPhaseCount                 = `sum by (phase) (kubevirt_vmi_phase_count{})`
-	avgVirtAPICPUUsage            = `avg(sum(rate(container_cpu_usage_seconds_total{namespace="kubevirt",pod=~"virt-api.*", container!="",container!="POD"}[%ds]))by (pod))`
+	avgVirtAPICPUUsage            = `avg(rate(container_cpu_usage_seconds_total{namespace="kubevirt",pod=~"virt-api.*", container!="",container!="POD"}[%ds]))`
```
Comment (alaypatel07) [inline tools/perfscale-audit/metric-client/metric-client.go:53]: the avgVirtAPIMemUsageInMB uses `avg(avg_over_time())` why does this not use the same `avg(avg_over_time(rate()))`?

File: tools/perfscale-audit/metric-client/metric-client.go [hunk 3]
```diff
@@ -50,15 +50,15 @@ const (
...
+	maxVirtAPIMemUsageInMB        = `avg(max_over_time(container_memory_rss{pod=~"virt-api.*", container!="POD", container!=""}[%ds]))/1024/1024`
 	avgVirtControllerMemUsageInMB = `max(avg_over_time(container_memory_rss{pod=~"virt-controller.*", container!="POD", container!=""}[%ds]))/1024/1024`
 	//  Finding the max value to get only the leader virt-controller pod data
-	minVirtControllerMemUsageInMB = `max(min by (pod)(min_over_time(container_memory_rss{pod=~"virt-controller.*", container!="POD", container!=""}[%ds])/1024/1024))`
-	maxVirtControllerMemUsageInMB = `max(max by (pod)(max_over_time(container_memory_rss{pod=~"virt-controller.*", container!="POD", container!=""}[%ds])/1024/1024))`
+	minVirtControllerMemUsageInMB = `max(min_over_time(container_memory_rss{pod=~"virt-controller.*", container!="POD", container!=""}[%ds]))/1024/1024`
+	maxVirtControllerMemUsageInMB = `max(max_over_time(container_memory_rss{pod=~"virt-controller.*", container!="POD", container!=""}[%ds]))/1024/1024`
+	virtControllerCPUUsage        = `max(rate(container_cpu_usage_seconds_total{namespace="kubevirt",pod=~"virt-controller.*", container!="",container!="POD"}[%ds]))`
```
Comment (alaypatel07) [inline tools/perfscale-audit/metric-client/metric-client.go:61]: same question as above, why dont we need `max(avg_over_time(rate()))`?

also another nit:s/virtControllerCPUUsage/avgVirtControllerCPUUsage

File: tools/perfscale-audit/metric-client/metric-client.go (hunk 2 above)
Comment (Sreeja1725) [reply tools/perfscale-audit/metric-client/metric-client.go:53]: container_memory_rss is a gauge metric, it is giving the memory usage at a given point in time. rate() is used in cases of counter metrics where data is accumulated over time. avg(avg_over_time()) is more appropriate in this case.

**Comment by alaypatel07** [comment]
/retitle update promql query of cpu and memory metrics for sig-performance tests

**Comment by Sreeja1725** [comment]
@alaypatel07  Updated the description and title

**Comment by alaypatel07** [comment]
/pull-kubevirt-e2e-k8s-1.29-sig-performance

**Comment by alaypatel07** [comment]
/lgtm

@rthallisey @xpivarc PTAL, this seems to be ready

File: tools/perfscale-audit/metric-client/metric-client.go [hunk 4]
```diff
@@ -50,15 +50,15 @@ const (
 //	at all times.
 const (
 	vmiPhaseCount                 = `sum by (phase) (kubevirt_vmi_phase_count{})`
-	avgVirtAPICPUUsage            = `avg(sum(rate(container_cpu_usage_seconds_total{namespace="kubevirt",pod=~"virt-api.*", container!="",container!="POD"}[%ds]))by (pod))`
+	avgVirtAPICPUUsage            = `avg(rate(container_cpu_usage_seconds_total{namespace="kubevirt",pod=~"virt-api.*", container!="",container!="POD"}[%ds]))`
 	avgVirtAPIMemUsageInMB        = `avg(avg_over_time(container_memory_rss{pod=~"virt-api.*", container!="POD", container!=""}[%ds]))/1024/1024`
-	minVirtAPIMemUsageInMB        = `min_over_time(container_memory_rss{pod=~"virt-api.*", container!="POD", container!=""}[%ds])/1024/1024`
-	maxVirtAPIMemUsageInMB        = `max_over_time(container_memory_rss{pod=~"virt-api.*", container!="POD", container!=""}[%ds])/1024/1024`
-	avgVirtControllerCPUUsage     = `avg(sum(rate(container_cpu_usage_seconds_total{namespace="kubevirt",pod=~"virt-controller.*", container!="",container!="POD"}[%ds])) by (pod))`
+	minVirtAPIMemUsageInMB        = `avg(min_over_time(container_memory_rss{pod=~"virt-api.*", container!="POD", container!=""}[%ds]))/1024/1024`
```
Comment (rthallisey) [inline tools/perfscale-audit/metric-client/metric-client.go:55]: We probably want to tell people we are taking an average.  I think `minVirtAPIMemUsageInMB` should be `avgMinVirtAPIMemUsageInMB`?  Same for `maxVirtAPIMemUsageInMB`.

**Review by rthallisey — CHANGES_REQUESTED** [review]

**Pull request merged**